carma_wm/src/CARMAWorldModel.cpp

```diff
@@ -0,0 +1,644 @@
+          continue;
+        }
+        else
+        {  // i == 1
+          first_2d = mutableCenterLine[i - 1];
+          second_2d = mutableCenterLine[i];
+          third_2d = mutableCenterLine[i + 1];
+          std::get<1>(vec.back()).push_back(computeCurvature(first_2d, second_2d, third_2d));
+          continue;
+        }
+      }
+      else if (n != 0 && i <= 1)
+      {
+        bool new_segment_needed = lanelet::geometry::distance2d(prevEndPoint, mutableCenterLine[i]) > 0.1;
+        if (new_segment_needed)
+        {
+          if (i == 0)
+          {
+            // Consider as disjoint lanelet
+            std::vector<double> curvatures;
+            vec.push_back(std::make_tuple(n, curvatures));
+            continue;
+          }
+          else
+          {  // i == 1
+            first_2d = mutableCenterLine[i - 1];
+            second_2d = mutableCenterLine[i];
+            third_2d = mutableCenterLine[i + 1];
+            std::get<1>(vec.back()).push_back(computeCurvature(first_2d, second_2d, third_2d));
+            continue;
+          }
+        }
+      }
+
+      first_2d = second_2d;
+      second_2d = third_2d;
+      third_2d = mutableCenterLine[i + 1];
+      std::get<1>(vec.back()).push_back(computeCurvature(first_2d, second_2d, third_2d));
+    }
+
+    prevEndPoint = mutableCenterLine.back();
+  }
+
+  return vec;
+}
+
+lanelet::LaneletMapConstPtr CARMAWorldModel::getMap() const
+{
+  return std::static_pointer_cast<lanelet::LaneletMap const>(semantic_map_);  // Cast pointer to const variant
+}
+
+LaneletRouteConstPtr CARMAWorldModel::getRoute() const
+{
+  return std::static_pointer_cast<const lanelet::routing::Route>(route_);  // Cast pointer to const variant
+}
+
+void CARMAWorldModel::setMap(lanelet::LaneletMapPtr map)
+{
+  semantic_map_ = map;
+  // Build routing graph from map
+  lanelet::traffic_rules::TrafficRulesUPtr traffic_rules = lanelet::traffic_rules::TrafficRulesFactory::create(
+      lanelet::Locations::Germany, lanelet::Participants::VehicleCar);
+  lanelet::routing::RoutingGraphUPtr map_graph = lanelet::routing::RoutingGraph::build(*semantic_map_, *traffic_rules);
+  map_routing_graph_ = std::move(map_graph);
+}
+
+void CARMAWorldModel::setRoute(LaneletRoutePtr route)
+{
+  route_ = route;
+
+  lanelet::ConstLanelets path_lanelets(route_->shortestPath().begin(), route_->shortestPath().end());
+
+  shortest_path_view_ = lanelet::utils::createConstMap(path_lanelets, {});
+
+  computeDowntrackReferenceLine();
+}
+
+lanelet::LineString3d CARMAWorldModel::copyConstructLineString(const lanelet::ConstLineString3d& line) const
+{
+  std::vector<lanelet::Point3d> coppied_points;
+  coppied_points.reserve(line.size());
+
+  for (auto basic_p : line.basicLineString())
+  {
+    coppied_points.push_back(lanelet::Point3d(lanelet::utils::getId(), basic_p));
+  }
+
+  return lanelet::LineString3d(lanelet::utils::getId(), coppied_points);
+}
+
+void CARMAWorldModel::computeDowntrackReferenceLine()
```

Comment:
Just noting our discussion about the discontinuity in crosstrack whenever a lane change happens along the reference line, worth thinking of ways we can keep that in mind but I think your documentation is sufficient for now.